agent/src/main/scala/za/co/absa/atum/agent/model/Measurement.scala

```diff
@@ -16,81 +16,26 @@
 
 package za.co.absa.atum.agent.model
 
-import za.co.absa.atum.agent.exception.AtumAgentException.MeasurementProvidedException
-import za.co.absa.atum.model.dto.MeasureResultDTO.ResultValueType
+import za.co.absa.atum.agent.exception.AtumAgentException.MeasurementException
+import za.co.absa.atum.agent.model.Measurement.validateMeasurement
 
 /**
- *  This trait defines a contract for a measurement.
+ *  This class defines a contract for a measurement.
  */
-trait Measurement {
-  val measure: Measure
-  val resultValue: Any
-  val resultType: ResultValueType.ResultValueType
+final case class Measurement private (measure: AtumMeasure, result: MeasureResult) {
+  validateMeasurement(measure, result)
 }
 
-/**
- *  This object contains all the possible measurements
- */
 object Measurement {
 
-  /**
-   *  When the application/user of Atum Agent provides actual results by himself, the type is precise and we don't need
-   *  to do any adjustments.
-   */
-  case class MeasurementProvided[T](measure: Measure, resultValue: T, resultType: ResultValueType.ResultValueType)
-      extends Measurement
-
-  object MeasurementProvided {
-
-    private def handleSpecificType[T](
-      measure: AtumMeasure,
-      resultValue: T,
-      requiredType: ResultValueType.ResultValueType
-    ): MeasurementProvided[T] = {
+  private def validateMeasurement(measure: AtumMeasure, result: MeasureResult): Unit = {
+    val actualType = result.resultValueType
+    val requiredType = measure.resultValueType
```

Comment:
This lead to runtime error. Would prefer to ensure it cannot happen. Isn't that possible?

Reply:
It absolutely is possible I believe - but the changes we would need to implement all begin with the measurement result value being `T` instead of `Any` - https://github.com/AbsaOSS/atum-service/pull/104#discussion_r1467322707, and I decided not to do it here.

The second step can be to use something like Shapeless and implement a combination of all measures and their data types that they support (I mean, just an implicit functions, based on example Filip provided)